src/Data/Singletons.hs

```diff
@@ -188,3 +211,69 @@ singThat p = withSing $ \x -> if p (fromSing x) then Just x else Nothing
 singByProxy :: SingI a => proxy a -> Sing a
 singByProxy _ = sing
 
+-- -------------------------------------------------------------------------- --
+-- Constraint Dictionaries
+
+data Dict :: Constraint -> Type -> Type where
+    Dict :: c => a -> Dict c a
+
+-- -------------------------------------------------------------------------- --
+-- Type Level Peano Numbers
+
+data N = Z | S N
```

Comment:
Surely we don't need to write this. Isn't there a Richard library somewhere for this, aka `singletons` or something? @emilypi

Reply:
I think these are already defined in `GHC.TypeNats` - is there a reason we redefined it here?

Reply:
The singletons package is highly experimental. It used to be bound to only the latest versions of GHC. The API broke regularly. It also used to be very big, since it included a lifted version of the complete base package.

I think, some of this may have changed with version 3.0 which was release only about a month ago (2021-03-12). I haven't yet gotten a chance to take a closer look. The code in this module comes from a draft for a micro-singletons package. I think singletons-3.0 may eliminate the need for such a package. For now, I think, it's too early to make a call on that. It would also be work for a separate PR.

Reply:
@emilypi GHC.TypeNats does not implement Peano numbers. It only provides type level literals. Unfortunately, the literals from that module can not be used for inductive proofs.

There are some packages on Hackage that work around that limitation by either providing an inductive representation on the fly or by directly injecting evidence for theorems over naturals into the type checker. I tried a few of those in the past, but some require type checker plugins, other didn't seem stable, and some were buggy.

In lack of a commonly used and stable library for type level Peano numbers, it seems the easiest to just add a minimal set of definitions that suffice for our use case here.

Specifically, there is the singletons-nat package. But that package used to come with its own singletons infrastructure hasn't been compatible with the definitions in this module. The most recent version of singletons-nat is base on the new minimal singletons-3.0 package, which is worth taking a closer look. But I propose that we first evaluate how stable singletons-3.0 is going to be and then possibly migrate to it in a separate PR.

Reply:
The code in this module is mostly a minimal core of definitions from the singletons package at the time this module was written.

Yes, singletons-3.0 looks much leaner and mature than previous versions and I agree that we should consider taking it as a new dependency instead of this code.

The code we are talking here has been in chainweb since the very first version. So, let's keep it for a separate PR to update it. This PR is already too large.